### ISSUE TITLE
TpetraCore_Issue1454_MPI_4 test failing on new cuda 9.2 ATDM build on white/ride

### ISSUE THREAD

**Issue by fryeguy52** [trilinos/Trilinos#3289]
CC: @trilinos/tpetra , @kddevin  (Trilinos Data Services Product Lead) @bartlettroscoe
 
## Next Action Status

PR #3318 merged on 8/18/2018 and this test started passing in these `cuda-9.2` builds on 8/19/2018.

## Description

As shown in [this query](https://testing-vm.sandia.gov/cdash/queryTests.php?project=Trilinos&filtercombine=and&filtercombine=and&filtercombine=and&date=2018-08-11&filtercombine=and&filtercombine=and&filtercombine=and&filtercombine=and&filtercombine=and&filtercombine=and&filtercount=5&showfilters=1&filtercombine=and&field1=buildname&compare1=65&value1=Trilinos-atdm-white-ride-cuda-9.2-&field2=buildname&compare2=64&value2=Trilinos-atdm-white-ride-cuda-9.2-debug-pt&field3=status&compare3=62&value3=passed&field4=status&compare4=62&value4=notrun&field5=testname&compare5=63&value5=TpetraCore_) the tests:

* TpetraCore_Issue1454_MPI_4 

are failing in the builds:

* Trilinos-atdm-white-ride-cuda-9.2-opt
* Trilinos-atdm-white-ride-cuda-9.2-debug

The test is timing out in 10 minutes.  This test takes at most 30 seconds on other builds of the ATDM configuration.

Some of the test output:
```
[white27:41375] mca_base_component_repository_open: unable to open mca_coll_hcoll: libsharp_coll.so.2: cannot open shared object file: No such file or directory (ignored)
[white27:41376] mca_base_component_repository_open: unable to open mca_coll_hcoll: libsharp_coll.so.2: cannot open shared object file: No such file or directory (ignored)
[white27:41378] mca_base_component_repository_open: unable to open mca_coll_hcoll: libsharp_coll.so.2: cannot open shared object file: No such file or directory (ignored)
[white27:41377] mca_base_component_repository_open: unable to open mca_coll_hcoll: libsharp_coll.so.2: cannot open shared object file: No such file or directory (ignored)
Teuchos::GlobalMPISession::GlobalMPISession(): started processor with name white27 and rank 0!
Teuchos::GlobalMPISession::GlobalMPISession(): started processor with name white27 and rank 1!
Teuchos::GlobalMPISession::GlobalMPISession(): started processor with name white27 and rank 2!
Teuchos::GlobalMPISession::GlobalMPISession(): started processor with name white27 and rank 3!

***
*** Unit test suite ...
***


Sorting tests by group name then by the order they were added ... (time = 2.18e-07)

Running unit tests ...

0. Distributor_Issue1454_UnitTest ... [white27:41378:0] Caught signal 11 (Segmentation fault)
[white27:41375:0] Caught signal 11 (Segmentation fault)
[white27:41376:0] Caught signal 11 (Segmentation fault)
[white27:41377:0] Caught signal 11 (Segmentation fault)
```

## Steps to Reproduce

One should be able to reproduce this failure on the machine `white` as described in:

* https://github.com/trilinos/Trilinos/blob/develop/cmake/std/atdm/README.md

More specifically, the commands given for the system `white` are provided at:

* https://github.com/trilinos/Trilinos/blob/develop/cmake/std/atdm/README.md#ridewhite

The exact commands to reproduce this issue should be:

```
$ cd <some_build_dir>/

$ source $TRILINOS_DIR/cmake/std/atdm/load-env.sh cuda-9.2-opt

$ cmake \
 -GNinja \
 -DTrilinos_CONFIGURE_OPTIONS_FILE:STRING=cmake/std/atdm/ATDMDevEnv.cmake \
 -DTrilinos_ENABLE_TESTS=ON -DTrilinos_ENABLE_Tpetra=ON \
 $TRILINOS_DIR

$ make NP=16

$ bsub -x -Is -q rhel7F -n 16 ctest -j16
```

**Comment by mhoemmen** [comment]
It looks like the test implicitly, inadvertently assumes that MPI can access UVM buffers.  `proc_ids_host` is a host mirror of a `Map<>::device_type` View, which happens to be a UVM View.  Host mirror of a UVM View is a UVM View.

**Comment by mhoemmen** [comment]
I'm a little busy today -- any @trilinos/tpetra folks want to take this?  The fix is to call the overload of `create_mirror_view` that takes a memory space as the first argument, and insist on that space being `HostSpace`.

**Comment by crtrott** [comment]
No that is not the fix: the fix is to require that MPI can access CUDA memory: because I think that what we said we would ask for as a baseline for Tpetra didn't we? Won't we otherwise always circumvent MPI optimizations for GPU-GPU transfers such as GPU-Direct?

**Comment by mhoemmen** [comment]
@crtrott The test actually does exercise access to a CudaSpace buffer.  The issue appears to be that the code passes a `CudaUVMSpace` allocation into a `Teuchos::ArrayView`, and hands that off to Distributor.

**Comment by crtrott** [comment]
And why would that be a problem?

**Comment by crtrott** [comment]
I mean what is the technical way of how this becomes an issue, since the UVM memory is supposed to work with access from host.

**Comment by mhoemmen** [comment]
@crtrott wrote:
> And why would that be a problem?

It's potentially slow?  I don't know, I agree that we should configure MPI correctly.  @tjfulle is working on this.

**Comment by bartlettroscoe** [comment]
FYI @nmhamster

**Comment by tjfulle** [comment]
I've got Trilinos building on White so I can look in to this.  I have a few other things to do, so I may not finish till tomorrow morning.

**Comment by mhoemmen** [comment]
Thanks @tjfulle ! :D

**Comment by tjfulle** [comment]
@fryeguy52, why do you say the test times out, it looks like it fails from a segmentation fault?

@mhoemmen the test fails in the call to `distributor.doPostsAndWaits` on line 141.  Is this related to `create_mirror_view` as you described above?  Using the overload that takes a memory space (and using the host memory space) does not fix the segmentation fault.

**Comment by bartlettroscoe** [comment]
@mhoemmen said:

> @fryeguy52, why do you say the test times out, it looks like it fails from a segmentation fault?

CTest was reporting this failure as a timeout due to the option `--mca orte_abort_on_non_zero_status 0` being passed into `srun`.  That is why it was listed as a Timeout.  But since the merge of #3292, now the test setgfaults and errors our right away at just under 4 sec as shown [here](https://testing-vm.sandia.gov/cdash/queryTests.php?project=Trilinos&filtercombine=and&filtercombine=and&filtercombine=and&filtercombine=and&filtercount=5&showfilters=1&filtercombine=and&field1=buildname&compare1=65&value1=Trilinos-atdm-white-ride-cuda-9.2&field2=buildname&compare2=62&value2=Trilinos-atdm-white-ride-cuda-9.2-debug-pt&field3=site&compare3=61&value3=white&field4=status&compare4=62&value4=passed&field5=status&compare5=62&value5=notrun).

**Comment by tjfulle** [comment]
@mhoemmen, this test fails somewhere in `Distributor::doPosts`.  I've put in some debugging code to track it down further.  Compile times on `white` are super long, so it is taking much longer than I would like.

**Comment by bartlettroscoe** [comment]
FYI: I sent and email to the white and ride admins just now with the below body.

Hopefully they can get this fixed and update the env.  (And hopefully we can test the ATDM builds of Trilinos and catch and resolve issues like this before and not after an update of these systems.)

---

Hello White and Ride Admins,

It seems from:

   https://github.com/trilinos/Trilinos/issues/3289

that the MPI installed as part of the CUDA 9.2 env may not be configured correctly to work directly with CUDA/GPU memory.  With the MPI with the old CUDA 8.0 on ‘white’ and ‘ride’, this was not a problem.

For more information, please confer with Si Hammond, Christian Trott, and Mark Hoemmen (all CCed).

Thanks,

-Ross

**Comment by tjfulle** [comment]
I've isolated the failure to a call to `Teuchos::send<T>` in `Distributor::doPosts`.  I can't say for certain it is an MPI problem, but it appears that way.  Compile times are so long on `white` that I am going to suspend further work on this issue until it is confirmed that MPI is configured correctly.

**Comment by mhoemmen** [comment]
Thanks @tjfulle for looking at this!  Is that the test that is trying to `send` UVM memory?

**Comment by tjfulle** [comment]
@mhoemmen I believe so

**Comment by mhoemmen** [comment]
@tjfulle I think we should just fix the test in the following way:

  1. Don't ever send CudaUVMSpace buffers to MPI
  2. Use Behavior to decide whether to send CudaSpace buffers to MPI

That shouldn't be too hard.  If (2) still breaks, then the MPI implementation is totally broken.

**Comment by mhoemmen** [comment]
@bartlettroscoe If this hinders work, please feel free to disable it.  I would rather get to the bottom of this, though.

**Comment by bartlettroscoe** [comment]
@mhoemmen said:

> If this hinders work, please feel free to disable it. I would rather get to the bottom of this, though.

How much longer do you think it will take to track this down?  I want to set up a process where we can allow tests to fail for a while and still keep on top of things and be able to see new failures. This way, when something gets fixed, we will just see it.  That is what we will work toward as part of #2933.

**Comment by mhoemmen** [comment]
@bartlettroscoe I'm in a hackathon right now, but I'll take a break from it and change the test so it doesn't do UVM buffer access.

**Comment by mhoemmen** [comment]
@bartlettroscoe PR https://github.com/trilinos/Trilinos/pull/3318 attempts to fix this issue.  Could you please review?  Thanks!

**Comment by tjfulle** [comment]
Thanks @mhoemmen , I got swamped with other things and was circling back just now to this.  Might we move some of the checks in the test to `doPostsAndWaits` itself to avoid this type of failure originating in other codes?

**Comment by mhoemmen** [comment]
@tjfulle wrote:
> I got swamped with other things and was circling back just now to this.
No worries! :D 

> Might we move some of the checks in the test to `doPostsAndWaits` itself...?

The `doPostsAndWaits` overload that takes `Kokkos::View` already uses a `static_assert` to exclude `CudaUVMSpace`.  The issue appears to be the `doPostsAndWaits` overload that takes `Teuchos::ArrayView`.  ArrayView can store arbitrary pointers and does not check whether they are CUDA allocations.  Distributor could use the the CUDA run-time API function that detects whether a pointer points to a device allocation.  However, my view (ha! ;-P ) is that users shouldn't be using host data structures to pass around pointers they can't access on host.  If we don't have truly unified, symmetric UVM, then we shouldn't be managing UVM allocations with things like `Teuchos::ArrayView`.

**Comment by mhoemmen** [comment]
@bartlettroscoe I just merged PR #3318 -- thanks for the review!  I'm hoping that fixes things.

Tpetra queries the MPI implementation at configure time to see whether it might possibly know how to deal with CUDA device buffers.  You can also control this option with an environment variable.  If Tpetra is told not to assume that MPI can handle CUDA device buffers, then Tpetra will not attempt to do boundary exchange with CUDA device buffers.  We could always use that environment variable if necessary, but as @crtrott points out, we really need MPI + CUDA to work correctly.

**Comment by bartlettroscoe** [comment]
>  @crtrott points out, we really need MPI + CUDA to work correctly.

@mhoemmen, that is not something that I can control.  We have to use the envs we are given.

**Comment by mhoemmen** [comment]
@bartlettroscoe I know _you_ can't control it :-D  The aforementioned PR #3318 should be able to work around badly configured MPI implementations.  If not, then MPI is totally broken on those machines and we need to fix their MPI configuration before we worry about Trilinos issues there.  Thanks Ross!

**Comment by bartlettroscoe** [comment]
Looks like PR #3318 fixed this failing test as it was newly passing on 8/19/2018 at shown [here](https://testing-vm.sandia.gov/cdash/viewTest.php?onlydelta&buildid=3849263).

Closing as complete.  Thanks @mhoemmen!

**Comment by bartlettroscoe** [comment]
@mhoemmen, in the end, was this really a defect in CUDA UVM handing in OpenMPI 3.1.0 on 'waterman'?  Downgrading to OpenMPI 2.1.2 on 'waterman' fixed similar issues in other tests (see #3290).  If that is the case, we should add the label "ATDM Env Issue".

We are trying to get a handle on how many Trilinos test failures are due to problems with the env vs. real Trilinos code or test bugs.

**Comment by mhoemmen** [comment]
@bartlettroscoe I think it's OK to add the label.  In general, the MPI + CUDA experience on all these platforms has been disappointing, and Trilinos' fixes' have actually mostly been work-arounds for not-very-functional environments.
